### ISSUE TITLE
Try the NIMs and KServe guide

### ISSUE THREAD

**Issue by kimwnasptd** [canonical/bundle-kubeflow#1077]
### Context

We want to integrate Canonical's [NIMs guide with KServe](https://pages.ubuntu.com/rs/066-EOV-335/images/NVIDIA_NIM_on_Charmed_Kubeflow_Tutorial.pdf) as part of our Charmed Kubeflow documentation.

To do so we'll need to run and verify the steps of the above guide, and document any modifications that we need to make for users to use NIMs with CKF.

With a quick look we might need to reconsider:
1. Downloading the `ngc` CLI and downloading the model in the PVC (the NIM container should do this on first launch)
2. Manually create a `PersistentVolume`. A PVC should be enough, which will be used by the ISVCs
3. Instead of defining a custom `ClusterServingRuntime` and `InferenceService` we could use the [ones from NVIDIA](https://github.com/NVIDIA/nim-deploy/tree/main/kserve)

### What needs to get done

1. Run the tutorial
2. Write down deviations we could do

### Definition of Done

1. We have run the paper above
2. We have documented if we want some changes to the above paper
3. Are ready to go and handle the docs task

**Comment by NohaIhab** [comment]
* the microk8s, juju, and charmed kubeflow versions are outdated in the existing guide. In the new guide, the following versions shall be used:
  * microk8s 1.29-strict/stable
  * juju 3.4/stable
  * kubeflow bundle 1.9/stable
* For the section on installing microk8s, juju, and deploying CKF we can link our tutorial instead of duplicating the instructions there
* In CKF 1.9, there's no need for the step to `Enable the NodeSelector feature of KServe`, it's already enabled by default for the corresponding version of `knative-serving`. In the new guide, we should omit this step.

**Comment by mvlassis** [comment]
Great exploration Noha! 
- Regarding the Ubuntu version, is 22.04 OK, or should we consider moving it to 24.04?
- We should use `juju status` instead of the `watch` command as we do in most of our documentation
- I agree that for the actual part of the Kubeflow installation, users should be redirected to our other guides to avoid duplication
- I also agree that the `InferenceService` and `Secret` should be created in the user's namespace, and the guide should reflect that (with the disabling of the sidecar injection)
- We could add a task or a TODO once we upgrade kserve to `0.14`

**Comment by kimwnasptd** [comment]
Adding some notes also from the demo @NohaIhab did, 3 days ago:

1. let's try to expose the command `ngc registry model list`, to help users know how they can list available models
2. How can we know the model to pull from NGC based on the OCI image from the `ClusterRuntime`?
3. The PV creation is painful (needs labels for specific nodes) 
4. Would not be cloud agnostic if we need to copy files from node in the PV of the node

During the above demo we came with the proposal of using the Volumes web-app of Kubeflow to create a RWX PVC and then via the Notebook download the model etc. This way the whole flow will be handled from KF

**Comment by NohaIhab** [comment]
thanks for the feedback @mvlassis and @kimwnasptd 
working on it

**Comment by NohaIhab** [comment]
## NIMs from Kubeflow UI
To make the UX better, I tried the following based on the above suggestions:
1. From the host CLI, create a `ClusterServingRuntime` using [the yaml from NVIDIA](https://github.com/NVIDIA/nim-deploy/blob/main/kserve/runtimes/llama3-8b-instruct-1.0.0.yaml). This will be indicated a step done by admin Users in the guide.
2. From the Volumes UI, create a new Volume with size `100Gi` and access mode `ReadWriteMany`
3. Create a Notebook, and attach the volume created in step `1.` as a Data volume from the Notebook creation page
3. From inside the JupyerLab, open a new terminal and create the Secrets needed for NIMs:
  * `docker-registry` Secret to be used as `ImagePullSecret`
  * secret with the env variable `NGC_CLI_API_KEY` to be used as env variable in the ISVC
Note that the names for the above secrets have to be the same as those defined in the runtime
5. From the same terminal inside the JupyterLab, create the Inference Service using the [yaml](https://github.com/NVIDIA/nim-deploy/blob/main/kserve/nim-models/llama3-8b-instruct_1xgpu_1.0.0.yaml) from NVIDIA repo
Note that the pvc defined in the `storageUri` attribute has to be the same as the pvc attached to the Notebook as a data volume
6. Wait for the InferenceService pod to be ready. When it's ready the isvc pod logs are:
```
INFO 10-14 10:22:29.134 api_server.py:456] Serving endpoints:
  0.0.0.0:8000/openapi.json
  0.0.0.0:8000/docs
  0.0.0.0:8000/docs/oauth2-redirect
  0.0.0.0:8000/metrics
  0.0.0.0:8000/v1/health/ready
  0.0.0.0:8000/v1/health/live
  0.0.0.0:8000/v1/models
  0.0.0.0:8000/v1/version
  0.0.0.0:8000/v1/chat/completions
  0.0.0.0:8000/v1/completions
INFO 10-14 10:22:29.134 api_server.py:460] An example cURL request:
curl -X 'POST' \
  'http://0.0.0.0:8000/v1/chat/completions' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "meta/llama3-8b-instruct",
    "messages": [
      {
        "role":"user",
        "content":"Hello! How are you?"
      },
      {
        "role":"assistant",
        "content":"Hi! I am quite well, how can I help you today?"
      },
      {
        "role":"user",
        "content":"Can you write me a song?"
      }
    ],
    "top_p": 1,
    "n": 1,
    "max_tokens": 15,
    "stream": true,
    "frequency_penalty": 1.0,
    "stop": ["hello"]
  }'
```
7. Get the isvc URL and make a request to the ISVC
```
KSERVE=$(kubectl get inferenceservice llama3-8b-instruct-1xgpu -o jsonpath='{.status.address.url}')

curl $KSERVE/v1/chat/completions -H "Content-Type: application/json" -d '{
"model": "meta/llama3-8b-instruct",
"messages": [{"role":"user","content":"What is Kubeflow?"}]
}'
{"id":"cmpl-f3ff374ad1cc454c80bde63889fdcfe2","object":"chat.completion","created":1728903534,"model":"meta/llama3-8b-instruct","choices":[{"index":0,"message":{"role":"assistant","content":"Kubeflow is an open-source platform for Machine Learning (ML) on Kubernetes. It allows data scientists and developers to easily build, deploy, and manage machine learning workloads on Kubernetes, a container orchestration system. Kubeflow provides a set of tools and libraries that simplify the deployment and management of ML pipelines, experiments, and models on Kubernetes clusters.\n\nKubeflow was originally developed by Google and is now maintained by the Linux Foundation. It is designed to address the growing needs of machine learning teams by providing a scalable, manageable, and reproducible way to develop, train, and deploy ML models.\n\nThe Kubeflow platform provides several key features, including:\n\n1. **Kubeflow Pipelines**: A workflow management system that allows users to define, execute, and manage Machine Learning pipelines.\n2. **Kubeflow Model Server**: A scalable and manageable way to deploy and manage machine learning models.\n3. **Kubeflow TensorFlowIntegration**: A set of tools and libraries that provide a seamless integration between TensorFlow and Kubernetes.\n4. **Kubeflow Jupyter**: A managed Jupyter notebook environment that allows users to develop and experiment with ML models.\n5. **Kubeflow Apache Beam Integration**: A set of tools and libraries that enable the execution of Apache Beam pipelines on Kubernetes.\n\nKubeflow provides several benefits, including:\n\n* **Scalability**: Kubeflow can scale to large clusters and handle complex ML workloads.\n* **Manageability**: Kubeflow provides a centralized platform for managing ML workloads, making it easier to track and debug experiments.\n* **Reproducibility**: Kubeflow ensures reproducibility of ML results by providing a consistent environment for experimentation and deployment.\n* **Collaboration**: Kubeflow enables collaboration among data scientists and engineers by providing a shared platform for experimentation and model deployment.\n\nKubeflow is widely used in various industries and organizations, including Google, Microsoft, Amazon, Facebook, and many research institutions and startups."},"logprobs":null,"finish_reason":"stop","stop_reason":128009}],"usage":{"prompt_tokens":16,"total_tokens":426,"completion_tokens":410}}
```

**Comment by NohaIhab** [comment]
## Pre-download model Issue
I experimented with trying to populate the models cache before creating the InferenceService by doing the following steps:
1. Run the job to list the compatible profiles for the given NIM
```
apiVersion: batch/v1
kind: Job
metadata:
  name: nim-profile-job
spec:
  template:
    metadata:
      annotations:
        "sidecar.istio.io/inject": "false"
      name: nim-profile-pod
    spec:
      containers:
      - name: nim-profile
        # Update the image name to the NIM that will be deployed in production
        image: nvcr.io/nim/meta/llama3-8b-instruct:1.0.0
        args: ["list-model-profiles"]
        env:
        - name: NIM_CACHE_PATH
          value: /tmp
        - name: NGC_API_KEY
          valueFrom:
            secretKeyRef:
              name: ngc-nim-secret
              key: NGC_API_KEY    
      imagePullSecrets:
      - name: ngc-docker-secret
      restartPolicy: Never
```
2. Get the output from the list job
```
kubectl logs nim-profile-job-8wv4j

===========================================
== NVIDIA Inference Microservice LLM NIM ==
===========================================

NVIDIA Inference Microservice LLM NIM Version 1.0.0
Model: nim/meta/llama3-8b-instruct

Container image Copyright (c) 2016-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.

This NIM container is governed by the NVIDIA AI Product Agreement here:
https://www.nvidia.com/en-us/data-center/products/nvidia-ai-enterprise/eula/.
A copy of this license can be found under /opt/nim/LICENSE.

The use of this model is governed by the AI Foundation Models Community License
here: https://docs.nvidia.com/ai-foundation-models-community-license.pdf.

ADDITIONAL INFORMATION: Meta Llama 3 Community License, Built with Meta Llama 3. 
A copy of the Llama 3 license can be found under /opt/nim/MODEL_LICENSE.

SYSTEM INFO
- Free GPUs:
  -  [2237:10de] (0) NVIDIA A10G (A10G) [current utilization: 2%]
MODEL PROFILES
- Compatible with system and runnable:
  - c334b76d50783655bdf62b8138511456f7b23083553d310268d0d05f254c012b (tensorrt_llm-a10g-fp16-tp1-throughput)
  - 8835c31752fbc67ef658b20a9f78e056914fdef0660206d82f252d62fd96064d (vllm-fp16-tp1)
  - With LoRA support:
    - 8d3824f766182a754159e88ad5a0bd465b1b4cf69ecf80bd6d6833753e945740 (vllm-fp16-tp1-lora)
- Incompatible with system:
  - dcd85d5e877e954f26c4a7248cd3b98c489fbde5f1cf68b4af11d665fa55778e (tensorrt_llm-h100-fp8-tp2-latency)
  - f59d52b0715ee1ecf01e6759dea23655b93ed26b12e57126d9ec43b397ea2b87 (tensorrt_llm-l40s-fp8-tp2-latency)
  - 30b562864b5b1e3b236f7b6d6a0998efbed491e4917323d04590f715aa9897dc (tensorrt_llm-h100-fp8-tp1-throughput)
  - 09e2f8e68f78ce94bf79d15b40a21333cea5d09dbe01ede63f6c957f4fcfab7b (tensorrt_llm-l40s-fp8-tp1-throughput)
  - a93a1a6b72643f2b2ee5e80ef25904f4d3f942a87f8d32da9e617eeccfaae04c (tensorrt_llm-a100-fp16-tp2-latency)
  - e0f4a47844733eb57f9f9c3566432acb8d20482a1d06ec1c0d71ece448e21086 (tensorrt_llm-a10g-fp16-tp2-latency)
  - 879b05541189ce8f6323656b25b7dff1930faca2abe552431848e62b7e767080 (tensorrt_llm-h100-fp16-tp2-latency)
  - 24199f79a562b187c52e644489177b6a4eae0c9fdad6f7d0a8cb3677f5b1bc89 (tensorrt_llm-l40s-fp16-tp2-latency)
  - 751382df4272eafc83f541f364d61b35aed9cce8c7b0c869269cea5a366cd08c (tensorrt_llm-a100-fp16-tp1-throughput)
  - cb52cbc73a6a71392094380f920a3548f27c5fcc9dab02a98dc1bcb3be9cf8d1 (tensorrt_llm-h100-fp16-tp1-throughput)
  - d8dd8af82e0035d7ca50b994d85a3740dbd84ddb4ed330e30c509e041ba79f80 (tensorrt_llm-l40s-fp16-tp1-throughput)
  - 19031a45cf096b683c4d66fff2a072c0e164a24f19728a58771ebfc4c9ade44f (vllm-fp16-tp2)
  - 9137f4d51dadb93c6b5864a19fd7c035bf0b718f3e15ae9474233ebd6468c359 (tensorrt_llm-a10g-fp16-tp2-throughput-lora)
  - cce57ae50c3af15625c1668d5ac4ccbe82f40fa2e8379cc7b842cc6c976fd334 (tensorrt_llm-a100-fp16-tp1-throughput-lora)
  - 3bdf6456ff21c19d5c7cc37010790448a4be613a1fd12916655dfab5a0dd9b8e (tensorrt_llm-h100-fp16-tp1-throughput-lora)
  - 388140213ee9615e643bda09d85082a21f51622c07bde3d0811d7c6998873a0b (tensorrt_llm-l40s-fp16-tp1-throughput-lora)
  - c5ffce8f82de1ce607df62a4b983e29347908fb9274a0b7a24537d6ff8390eb9 (vllm-fp16-tp2-lora)
```
3. Run the download profile job by modifying the job yaml to download the profile from the list job output:
```
apiVersion: batch/v1
kind: Job
metadata:
  name: nim-download-profile-job
spec:
  template:
    metadata:
      annotations:
        "sidecar.istio.io/inject": "false"
      name: nim-download-profile-pod
    spec:
      containers:
      - name: nim-profile-single
        # Update the image name to the NIM that will be deployed in production
        image: nvcr.io/nim/meta/llama3-8b-instruct:1.0.0
        
        # Update the profile name to either the hash or the human-readable name that will be used in production
        args: ["download-to-cache", "--profile", "c334b76d50783655bdf62b8138511456f7b23083553d310268d0d05f254c012b"]
        env:
        - name: NIM_CACHE_PATH
          value: /mnt/models/cache
        - name: NGC_API_KEY
          valueFrom:
            secretKeyRef:
              name: ngc-nim-secret
              key: NGC_API_KEY    
        volumeMounts:
        - name: model-cache
          mountPath: /mnt/models
      imagePullSecrets:
      - name: ngc-docker-secret
      volumes:
      - name: model-cache
        persistentVolumeClaim:
          claimName: nims-vol
      restartPolicy: Never
```
Now we have the profile download in the `nims-vol` directory under `/cache`
4. Apply the llama3 runtime to the cluster with the yaml:
```
apiVersion: serving.kserve.io/v1alpha1
kind: ClusterServingRuntime
metadata:
  name: cached-nvidia-nim-llama3-8b-instruct-1.0.0
spec:
  annotations:
    prometheus.kserve.io/path: /metrics
    prometheus.kserve.io/port: "8000"
    serving.kserve.io/enable-metric-aggregation: "true"
    serving.kserve.io/enable-prometheus-scraping: "true"
  containers:
  - env:
    - name: NIM_CACHE_PATH
      value: /mnt/models/cache
    - name: NGC_API_KEY
      valueFrom:
        secretKeyRef:
          name: ngc-nim-secret
          key: NGC_API_KEY
    image: nvcr.io/nim/meta/llama3-8b-instruct:1.0.0
    name: kserve-container
    ports:
    - containerPort: 8000
      protocol: TCP
    resources:
      limits:
        cpu: "1"
        memory: 16Gi
      requests:
        cpu: "1"
        memory: 16Gi
    volumeMounts:
    - mountPath: /dev/shm
      name: dshm
  imagePullSecrets:
  - name: ngc-docker-secret
  protocolVersions:
  - v2
  - grpc-v2
  supportedModelFormats:
  - autoSelect: true
    name: nvidia-nim-llama3-8b-instruct
    priority: 1
    version: 1.0.0
  volumes:
  - emptyDir:
      medium: Memory
      sizeLimit: 16Gi
    name: dshm
```
5. Create the InferenceService and have it use the populated PVC as `storageUri`. The ISVC yaml is:
```
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  annotations:
    autoscaling.knative.dev/target: "10"
    "sidecar.istio.io/inject": "false"
  name: llama3-8b-instruct-1xgpu
spec:
  predictor:
    minReplicas: 1
    model:
      modelFormat:
        name: nvidia-nim-llama3-8b-instruct
      resources:
        limits:
          nvidia.com/gpu: "1"
        requests:
          nvidia.com/gpu: "1"
      runtime: cached-nvidia-nim-llama3-8b-instruct-1.0.0
      storageUri: pvc://nims-vol/
```
6. Get the logs from the InferenceService Pod
```
kubectl logs -f llama3-8b-instruct-1xgpu-predictor-00001-deployment-79c5f788xpg
Defaulted container "kserve-container" out of: kserve-container, queue-proxy

===========================================
== NVIDIA Inference Microservice LLM NIM ==
===========================================

NVIDIA Inference Microservice LLM NIM Version 1.0.0
Model: nim/meta/llama3-8b-instruct

Container image Copyright (c) 2016-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.

This NIM container is governed by the NVIDIA AI Product Agreement here:
https://www.nvidia.com/en-us/data-center/products/nvidia-ai-enterprise/eula/.
A copy of this license can be found under /opt/nim/LICENSE.

The use of this model is governed by the AI Foundation Models Community License
here: https://docs.nvidia.com/ai-foundation-models-community-license.pdf.

ADDITIONAL INFORMATION: Meta Llama 3 Community License, Built with Meta Llama 3. 
A copy of the Llama 3 license can be found under /opt/nim/MODEL_LICENSE.
WARNING 10-15 12:41:23.21 caches.py:27] /mnt/models/cache is read-only, application may fail if model is not already present in cache

2024-10-15 12:41:27,250 [INFO] PyTorch version 2.2.2 available.
2024-10-15 12:41:27,805 [WARNING] [TRT-LLM] [W] Logger level already set from environment. Discard new verbosity: error
2024-10-15 12:41:27,805 [INFO] [TRT-LLM] [I] Starting TensorRT-LLM init.
[TensorRT-LLM] TensorRT-LLM version: 0.10.1.dev2024053000
2024-10-15 12:41:27,916 [INFO] [TRT-LLM] [I] TensorRT-LLM inited.
INFO 10-15 12:41:28.770 api_server.py:489] NIM LLM API version 1.0.0
INFO 10-15 12:41:28.771 ngc_profile.py:217] Running NIM without LoRA. Only looking for compatible profiles that do not support LoRA.
INFO 10-15 12:41:28.771 ngc_profile.py:219] Detected 2 compatible profile(s).
INFO 10-15 12:41:28.772 ngc_injector.py:106] Valid profile: c334b76d50783655bdf62b8138511456f7b23083553d310268d0d05f254c012b (tensorrt_llm-a10g-fp16-tp1-throughput) on GPUs [0]
INFO 10-15 12:41:28.772 ngc_injector.py:106] Valid profile: 8835c31752fbc67ef658b20a9f78e056914fdef0660206d82f252d62fd96064d (vllm-fp16-tp1) on GPUs [0]
INFO 10-15 12:41:28.772 ngc_injector.py:141] Selected profile: c334b76d50783655bdf62b8138511456f7b23083553d310268d0d05f254c012b (tensorrt_llm-a10g-fp16-tp1-throughput)
INFO 10-15 12:41:30.265 ngc_injector.py:146] Profile metadata: precision: fp16
INFO 10-15 12:41:30.265 ngc_injector.py:146] Profile metadata: pp: 1
INFO 10-15 12:41:30.265 ngc_injector.py:146] Profile metadata: gpu_device: 2237:10de
INFO 10-15 12:41:30.265 ngc_injector.py:146] Profile metadata: profile: throughput
INFO 10-15 12:41:30.265 ngc_injector.py:146] Profile metadata: llm_engine: tensorrt_llm
INFO 10-15 12:41:30.265 ngc_injector.py:146] Profile metadata: gpu: A10G
INFO 10-15 12:41:30.265 ngc_injector.py:146] Profile metadata: feat_lora: false
INFO 10-15 12:41:30.265 ngc_injector.py:146] Profile metadata: tp: 1
INFO 10-15 12:41:30.265 ngc_injector.py:166] Preparing model workspace. This step might download additional files to run the model.
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.10/dist-packages/vllm_nvext/entrypoints/openai/api_server.py", line 492, in <module>
    engine_args, extracted_name = inject_ngc_hub(engine_args)
  File "/usr/local/lib/python3.10/dist-packages/vllm_nvext/hub/ngc_injector.py", line 168, in inject_ngc_hub
    cached = repo.download_all()
Exception: "Errors fetching files: \nissue a download request for file: \"tokenizer_config.json\"; the repo file list has not been validated. \nissue a download request for file: \"nvidia-smi.log\"; the repo file list has not been validated. \nissue a download request for file: \"model.safetensors.index.json\"; the repo file list has not been validated. \nissue a download request for file: \"rank0.engine\"; the repo file list has not been validated. \nissue a download request for file: \"trt_llm_config.yaml\"; the repo file list has not been validated. \nissue a download request for file: \"generation_config.json\"; the repo file list has not been validated. \nissue a download request for file: \"config.json\"; the repo file list has not been validated. \nissue a download request for file: \"tokenizer.json\"; the repo file list has not been validated. \nissue a download request for file: \"special_tokens_map.json\"; the repo file list has not been validated. \nissue a download request for file: \"checksums.blake3\"; the repo file list has not been validated. \nissue a download request for file: \"config.json\"; the repo file list has not been validated. \nissue a download request for file: \"metadata.json\"; the repo file list has not been validated. \nissue a download request for file: \"nim_config.yaml\"; the repo file list has not been validated. "
```
the `kserve-container` container constantly fails with the above error and goes to `CrashLoopBackOff` status

**Comment by NohaIhab** [comment]
## Run InferenceService without pre-downloading the model
It's possible to run the NIMs ISVC without pre-downloading the model into the PV by editing the runtime as follows:
```
@@ -12,5 +12,5 @@
   - env:
     - name: NIM_CACHE_PATH
-      value: /mnt/models/cache
+      value: /tmp
     - name: NGC_API_KEY
       valueFrom:
```
this tells the container to download the model in the memory, completely ignoring the PVC

This option is also documented in the `nim-deploy/kserve` repo under [Setup](https://github.com/NVIDIA/nim-deploy/blob/main/kserve/README.md#setup) search for `/tmp`

By making this change, the ISVC Pod is running successfully, with the following logs from `kserve-container`:
```
kubectl logs llama3-8b-instruct-1xgpu-predictor-00001-deployment-7bfc8824hjf
Defaulted container "kserve-container" out of: kserve-container, queue-proxy

===========================================
== NVIDIA Inference Microservice LLM NIM ==
===========================================

NVIDIA Inference Microservice LLM NIM Version 1.0.0
Model: nim/meta/llama3-8b-instruct

Container image Copyright (c) 2016-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.

This NIM container is governed by the NVIDIA AI Product Agreement here:
https://www.nvidia.com/en-us/data-center/products/nvidia-ai-enterprise/eula/.
A copy of this license can be found under /opt/nim/LICENSE.

The use of this model is governed by the AI Foundation Models Community License
here: https://docs.nvidia.com/ai-foundation-models-community-license.pdf.

ADDITIONAL INFORMATION: Meta Llama 3 Community License, Built with Meta Llama 3. 
A copy of the Llama 3 license can be found under /opt/nim/MODEL_LICENSE.

2024-10-15 13:17:39,919 [INFO] PyTorch version 2.2.2 available.
2024-10-15 13:17:40,551 [WARNING] [TRT-LLM] [W] Logger level already set from environment. Discard new verbosity: error
2024-10-15 13:17:40,551 [INFO] [TRT-LLM] [I] Starting TensorRT-LLM init.
2024-10-15 13:17:40,665 [INFO] [TRT-LLM] [I] TensorRT-LLM inited.
[TensorRT-LLM] TensorRT-LLM version: 0.10.1.dev2024053000
INFO 10-15 13:17:41.573 api_server.py:489] NIM LLM API version 1.0.0
INFO 10-15 13:17:41.574 ngc_profile.py:217] Running NIM without LoRA. Only looking for compatible profiles that do not support LoRA.
INFO 10-15 13:17:41.574 ngc_profile.py:219] Detected 2 compatible profile(s).
INFO 10-15 13:17:41.574 ngc_injector.py:106] Valid profile: c334b76d50783655bdf62b8138511456f7b23083553d310268d0d05f254c012b (tensorrt_llm-a10g-fp16-tp1-throughput) on GPUs [0]
INFO 10-15 13:17:41.574 ngc_injector.py:106] Valid profile: 8835c31752fbc67ef658b20a9f78e056914fdef0660206d82f252d62fd96064d (vllm-fp16-tp1) on GPUs [0]
INFO 10-15 13:17:41.574 ngc_injector.py:141] Selected profile: c334b76d50783655bdf62b8138511456f7b23083553d310268d0d05f254c012b (tensorrt_llm-a10g-fp16-tp1-throughput)
INFO 10-15 13:17:43.74 ngc_injector.py:146] Profile metadata: precision: fp16
INFO 10-15 13:17:43.74 ngc_injector.py:146] Profile metadata: gpu_device: 2237:10de
INFO 10-15 13:17:43.74 ngc_injector.py:146] Profile metadata: llm_engine: tensorrt_llm
INFO 10-15 13:17:43.74 ngc_injector.py:146] Profile metadata: pp: 1
INFO 10-15 13:17:43.74 ngc_injector.py:146] Profile metadata: gpu: A10G
INFO 10-15 13:17:43.74 ngc_injector.py:146] Profile metadata: feat_lora: false
INFO 10-15 13:17:43.74 ngc_injector.py:146] Profile metadata: profile: throughput
INFO 10-15 13:17:43.74 ngc_injector.py:146] Profile metadata: tp: 1
INFO 10-15 13:17:43.74 ngc_injector.py:166] Preparing model workspace. This step might download additional files to run the model.
INFO 10-15 13:18:53.457 ngc_injector.py:172] Model workspace is now ready. It took 70.383 seconds
INFO 10-15 13:18:53.461 async_trtllm_engine.py:74] Initializing an LLM engine (v1.0.0) with config: model='/tmp/meta--llama3-8b-instruct-7bmn0itv', speculative_config=None, tokenizer='/tmp/meta--llama3-8b-instruct-7bmn0itv', tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.bfloat16, max_seq_len=8192, download_dir=None, load_format=auto, tensor_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), seed=0)
WARNING 10-15 13:18:53.920 logging.py:314] Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
INFO 10-15 13:18:53.949 utils.py:201] Using 0 bytes of gpu memory for PEFT cache
INFO 10-15 13:18:53.949 utils.py:207] Engine size in bytes 16067779716
INFO 10-15 13:18:53.949 utils.py:211] available device memory 23606263808
INFO 10-15 13:18:53.949 utils.py:218] Setting free_gpu_memory_fraction to 0.9
WARNING 10-15 13:20:10.799 logging.py:314] Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
INFO 10-15 13:20:10.814 serving_chat.py:347] Using default chat template:
{% set loop_messages = messages %}{% for message in loop_messages %}{% set content = '<|start_header_id|>' + message['role'] + '<|end_header_id|>

'+ message['content'] | trim + '<|eot_id|>' %}{% if loop.index0 == 0 %}{% set content = bos_token + content %}{% endif %}{{ content }}{% endfor %}{% if add_generation_prompt %}{{ '<|start_header_id|>assistant<|end_header_id|>

' }}{% endif %}
WARNING 10-15 13:20:11.127 logging.py:314] Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
INFO 10-15 13:20:11.143 api_server.py:456] Serving endpoints:
  0.0.0.0:8000/openapi.json
  0.0.0.0:8000/docs
  0.0.0.0:8000/docs/oauth2-redirect
  0.0.0.0:8000/metrics
  0.0.0.0:8000/v1/health/ready
  0.0.0.0:8000/v1/health/live
  0.0.0.0:8000/v1/models
  0.0.0.0:8000/v1/version
  0.0.0.0:8000/v1/chat/completions
  0.0.0.0:8000/v1/completions
INFO 10-15 13:20:11.143 api_server.py:460] An example cURL request:
curl -X 'POST' \
  'http://0.0.0.0:8000/v1/chat/completions' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "meta/llama3-8b-instruct",
    "messages": [
      {
        "role":"user",
        "content":"Hello! How are you?"
      },
      {
        "role":"assistant",
        "content":"Hi! I am quite well, how can I help you today?"
      },
      {
        "role":"user",
        "content":"Can you write me a song?"
      }
    ],
    "top_p": 1,
    "n": 1,
    "max_tokens": 15,
    "stream": true,
    "frequency_penalty": 1.0,
    "stop": ["hello"]
  }'

INFO 10-15 13:20:11.194 server.py:82] Started server process [32]
INFO 10-15 13:20:11.195 on.py:48] Waiting for application startup.
INFO 10-15 13:20:11.197 on.py:62] Application startup complete.
INFO 10-15 13:20:11.198 server.py:214] Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
```
Since the runtime now does not rely on the PVC cache for getting the models, I experimented to omit the `storageUri` from the InferenceService yaml so it looks like this:
```
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  annotations:
    autoscaling.knative.dev/target: "10"
    "sidecar.istio.io/inject": "false"
  name: llama3-8b-instruct-1xgpu
spec:
  predictor:
    minReplicas: 1
    model:
      modelFormat:
        name: nvidia-nim-llama3-8b-instruct
      resources:
        limits:
          nvidia.com/gpu: "1"
        requests:
          nvidia.com/gpu: "1"
      runtime: nvidia-nim-llama3-8b-instruct-1.0.0
```
The result from running the InferenceService without the `storageUri`:
```

```

**Comment by NohaIhab** [comment]
## Conclusion
The guide will be written using the method described in https://github.com/canonical/bundle-kubeflow/issues/1077#issuecomment-2413918966.
Once we upgrade to KServe 0.14 where the issue https://github.com/kserve/kserve/issues/3687 is resolved i.e. Kubeflow 1.10, we should update the guide to create a pvc and have the NIMs container download the model in it the first time it runs.
I have it on my list to create an issue about updating the guide with the label Kubeflow 1.10 once the guide is created and published so I can link it there and specify the sections we will need to change.
